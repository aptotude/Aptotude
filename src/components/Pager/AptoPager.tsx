import React from 'react';
import classNames from 'classnames';
import { StandardTypes } from '../../utils/standardTypes';
import { AptoButton } from '../Button';
import './aptoPager.scss';

interface AptoPagerProps extends StandardTypes {
  current: number;
  total: number;
  onClick?: (page: number, event: React.MouseEvent) => void;
  linkCount?: number;
  pages?: boolean;
  nextPrev?: boolean;
}

const COMPONENT_PREFIX = 'AptoPager';

const getPagingRange = (
  current: number,
  { min = 1, total = 20, length = 5 } = {}
) => {
  if (length > total) {
    length = total;
  }
  let start = current - Math.floor(length / 2);
  start = Math.max(start, min);
  start = Math.min(start, min + total - length);
  return Array.from({ length }, (el, i) => start + i);
};

export class AptoPager extends React.Component<AptoPagerProps> {
  public static defaultProps = {
    pages: true,
    nextPrev: true,
    linkCount: 5
  };
  public clickHandler = (
    e: React.MouseEvent<HTMLButtonElement | HTMLAnchorElement>
  ) => {
    const { onClick } = this.props;
    if (onClick) {
      onClick(Number(e.currentTarget.dataset.page), e);
    }
  };

  public render() {
    const {
      pages,
      nextPrev,
      total,
      current,
      linkCount,
      className
    } = this.props;
    const classes = classNames(COMPONENT_PREFIX, className);
    const links: number[] = getPagingRange(current, {
      total,
      length: linkCount
    });
    const prev = current > 1 ? current - 1 : null;
    const next = current < total ? current + 1 : null;

    return (
      <div className={classes}>
        <ul>
          {prev && nextPrev && (
            <li className="AptoPager--item AptoPager--prev">
              <AptoButton
                kind="link"
                type="button"
                onClick={this.clickHandler}
                data-page={prev}
              >
                &lt; Previous
              </AptoButton>
            </li>
          )}
          {pages &&
            links.map(l => (
              <li className="AptoPager--item" key={`${l}`}>
                {(l === current && (
                  <AptoButton
                    kind="link"
                    type="button"
                    className="activePage"
                    onClick={this.clickHandler}
                    data-page={l}
                  >
                    {l}
                  </AptoButton>
                )) || (
                  <AptoButton
                    kind="link"
                    type="button"
                    onClick={this.clickHandler}
                    data-page={l}
                  >
                    {l}
                  </AptoButton>
                )}
              </li>
            ))}
          {next && nextPrev && (
            <li className="AptoPager--item AptoPager--next">
              <AptoButton
                kind="link"
                type="button"
                onClick={this.clickHandler}
                data-page={next}
              >
                Next &gt;
              </AptoButton>
            </li>
          )}
        </ul>
      </div>
    );
  }
}
