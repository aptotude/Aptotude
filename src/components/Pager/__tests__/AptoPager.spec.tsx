import * as React from 'react';
import {
  render,
  cleanup,
  fireEvent,
  getByText,
  wait
} from 'react-testing-library';
import { AptoPager } from '../AptoPager';

afterEach(() => {
  cleanup();
});

describe('AptoPager', () => {
  it('Renders Component', () => {
    const { container } = render(<AptoPager current={2} total={10} />);
    expect(container.firstChild).toMatchSnapshot();
  });

  it('Renders Component with classnames', () => {
    const { container } = render(
      <AptoPager className="fooo" current={2} total={10} />
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  it('Click event for pages', async () => {
    let clicked = 0;
    const { container } = render(
      <AptoPager
        onPageClick={pageNum => {
          clicked = pageNum;
        }}
        current={2}
        total={10}
      />
    );
    fireEvent.click(getByText(container, '3'));
    await wait(() => {
      expect(clicked).toEqual(3);
    });
  });

  it('Click event for prev', async () => {
    let clicked = 0;
    const { container } = render(
      <AptoPager
        onPageClick={pageNum => {
          clicked = pageNum;
        }}
        current={2}
        total={10}
      />
    );
    fireEvent.click(getByText(container, '< Previous'));
    await wait(() => {
      expect(clicked).toEqual(1);
    });
  });

  it('Click event for next', async () => {
    let clicked = 0;
    const { container } = render(
      <AptoPager
        onPageClick={pageNum => {
          clicked = pageNum;
        }}
        current={2}
        total={10}
      />
    );
    fireEvent.click(getByText(container, 'Next >'));
    await wait(() => {
      expect(clicked).toEqual(3);
    });
  });

  it('Click event does not fire on current page', async () => {
    let clicked = 0;
    const { container } = render(
      <AptoPager
        onPageClick={pageNum => {
          clicked = pageNum;
        }}
        current={2}
        total={10}
      />
    );
    fireEvent.click(getByText(container, '2'));
    await wait(() => {
      expect(clicked).toEqual(0);
    });
  });

  it('Renders no prev', () => {
    const { container } = render(<AptoPager current={1} total={10} />);
    expect(container.firstChild).toMatchSnapshot();
  });

  it('Renders no next', () => {
    const { container } = render(<AptoPager current={10} total={10} />);
    expect(container.firstChild).toMatchSnapshot();
  });

  it('Renders Component with more links', () => {
    const { container } = render(
      <AptoPager linkCount={10} current={2} total={100} />
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  it('Renders next/prev', () => {
    const { container } = render(
      <AptoPager pages={false} current={2} total={10} />
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  it('Renders pages', () => {
    const { container } = render(
      <AptoPager nextPrev={false} current={2} total={10} />
    );
    expect(container.firstChild).toMatchSnapshot();
  });
});
