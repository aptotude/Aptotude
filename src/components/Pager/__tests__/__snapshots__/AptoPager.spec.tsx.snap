// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AptoPager Renders Component 1`] = `
<div
  class="AptoPager"
>
  <ul>
    <li
      class="AptoPager--item AptoPager--prev"
    >
      <button
        class="AptoButton AptoButton--link AptoButton--primary"
        data-page="1"
        type="button"
      >
        &lt; Previous
      </button>
    </li>
    <li
      class="AptoPager--item"
    >
      <button
        class="AptoButton AptoButton--link AptoButton--primary"
        data-page="1"
        type="button"
      >
        1
      </button>
    </li>
    <li
      class="AptoPager--item"
    >
      <button
        class="AptoButton AptoButton--link AptoButton--primary activePage"
        data-page="2"
        type="button"
      >
        2
      </button>
    </li>
    <li
      class="AptoPager--item"
    >
      <button
        class="AptoButton AptoButton--link AptoButton--primary"
        data-page="3"
        type="button"
      >
        3
      </button>
    </li>
    <li
      class="AptoPager--item"
    >
      <button
        class="AptoButton AptoButton--link AptoButton--primary"
        data-page="4"
        type="button"
      >
        4
      </button>
    </li>
    <li
      class="AptoPager--item"
    >
      <button
        class="AptoButton AptoButton--link AptoButton--primary"
        data-page="5"
        type="button"
      >
        5
      </button>
    </li>
    <li
      class="AptoPager--item AptoPager--next"
    >
      <button
        class="AptoButton AptoButton--link AptoButton--primary"
        data-page="3"
        type="button"
      >
        Next &gt;
      </button>
    </li>
  </ul>
</div>
`;

exports[`AptoPager Renders Component with classnames 1`] = `
<div
  class="AptoPager fooo"
>
  <ul>
    <li
      class="AptoPager--item AptoPager--prev"
    >
      <button
        class="AptoButton AptoButton--link AptoButton--primary"
        data-page="1"
        type="button"
      >
        &lt; Previous
      </button>
    </li>
    <li
      class="AptoPager--item"
    >
      <button
        class="AptoButton AptoButton--link AptoButton--primary"
        data-page="1"
        type="button"
      >
        1
      </button>
    </li>
    <li
      class="AptoPager--item"
    >
      <button
        class="AptoButton AptoButton--link AptoButton--primary activePage"
        data-page="2"
        type="button"
      >
        2
      </button>
    </li>
    <li
      class="AptoPager--item"
    >
      <button
        class="AptoButton AptoButton--link AptoButton--primary"
        data-page="3"
        type="button"
      >
        3
      </button>
    </li>
    <li
      class="AptoPager--item"
    >
      <button
        class="AptoButton AptoButton--link AptoButton--primary"
        data-page="4"
        type="button"
      >
        4
      </button>
    </li>
    <li
      class="AptoPager--item"
    >
      <button
        class="AptoButton AptoButton--link AptoButton--primary"
        data-page="5"
        type="button"
      >
        5
      </button>
    </li>
    <li
      class="AptoPager--item AptoPager--next"
    >
      <button
        class="AptoButton AptoButton--link AptoButton--primary"
        data-page="3"
        type="button"
      >
        Next &gt;
      </button>
    </li>
  </ul>
</div>
`;

exports[`AptoPager Renders Component with more links 1`] = `
<div
  class="AptoPager"
>
  <ul>
    <li
      class="AptoPager--item AptoPager--prev"
    >
      <button
        class="AptoButton AptoButton--link AptoButton--primary"
        data-page="1"
        type="button"
      >
        &lt; Previous
      </button>
    </li>
    <li
      class="AptoPager--item"
    >
      <button
        class="AptoButton AptoButton--link AptoButton--primary"
        data-page="1"
        type="button"
      >
        1
      </button>
    </li>
    <li
      class="AptoPager--item"
    >
      <button
        class="AptoButton AptoButton--link AptoButton--primary activePage"
        data-page="2"
        type="button"
      >
        2
      </button>
    </li>
    <li
      class="AptoPager--item"
    >
      <button
        class="AptoButton AptoButton--link AptoButton--primary"
        data-page="3"
        type="button"
      >
        3
      </button>
    </li>
    <li
      class="AptoPager--item"
    >
      <button
        class="AptoButton AptoButton--link AptoButton--primary"
        data-page="4"
        type="button"
      >
        4
      </button>
    </li>
    <li
      class="AptoPager--item"
    >
      <button
        class="AptoButton AptoButton--link AptoButton--primary"
        data-page="5"
        type="button"
      >
        5
      </button>
    </li>
    <li
      class="AptoPager--item"
    >
      <button
        class="AptoButton AptoButton--link AptoButton--primary"
        data-page="6"
        type="button"
      >
        6
      </button>
    </li>
    <li
      class="AptoPager--item"
    >
      <button
        class="AptoButton AptoButton--link AptoButton--primary"
        data-page="7"
        type="button"
      >
        7
      </button>
    </li>
    <li
      class="AptoPager--item"
    >
      <button
        class="AptoButton AptoButton--link AptoButton--primary"
        data-page="8"
        type="button"
      >
        8
      </button>
    </li>
    <li
      class="AptoPager--item"
    >
      <button
        class="AptoButton AptoButton--link AptoButton--primary"
        data-page="9"
        type="button"
      >
        9
      </button>
    </li>
    <li
      class="AptoPager--item"
    >
      <button
        class="AptoButton AptoButton--link AptoButton--primary"
        data-page="10"
        type="button"
      >
        10
      </button>
    </li>
    <li
      class="AptoPager--item AptoPager--next"
    >
      <button
        class="AptoButton AptoButton--link AptoButton--primary"
        data-page="3"
        type="button"
      >
        Next &gt;
      </button>
    </li>
  </ul>
</div>
`;

exports[`AptoPager Renders next/prev 1`] = `
<div
  class="AptoPager"
>
  <ul>
    <li
      class="AptoPager--item AptoPager--prev"
    >
      <button
        class="AptoButton AptoButton--link AptoButton--primary"
        data-page="1"
        type="button"
      >
        &lt; Previous
      </button>
    </li>
    <li
      class="AptoPager--item AptoPager--next"
    >
      <button
        class="AptoButton AptoButton--link AptoButton--primary"
        data-page="3"
        type="button"
      >
        Next &gt;
      </button>
    </li>
  </ul>
</div>
`;

exports[`AptoPager Renders no next 1`] = `
<div
  class="AptoPager"
>
  <ul>
    <li
      class="AptoPager--item AptoPager--prev"
    >
      <button
        class="AptoButton AptoButton--link AptoButton--primary"
        data-page="9"
        type="button"
      >
        &lt; Previous
      </button>
    </li>
    <li
      class="AptoPager--item"
    >
      <button
        class="AptoButton AptoButton--link AptoButton--primary"
        data-page="6"
        type="button"
      >
        6
      </button>
    </li>
    <li
      class="AptoPager--item"
    >
      <button
        class="AptoButton AptoButton--link AptoButton--primary"
        data-page="7"
        type="button"
      >
        7
      </button>
    </li>
    <li
      class="AptoPager--item"
    >
      <button
        class="AptoButton AptoButton--link AptoButton--primary"
        data-page="8"
        type="button"
      >
        8
      </button>
    </li>
    <li
      class="AptoPager--item"
    >
      <button
        class="AptoButton AptoButton--link AptoButton--primary"
        data-page="9"
        type="button"
      >
        9
      </button>
    </li>
    <li
      class="AptoPager--item"
    >
      <button
        class="AptoButton AptoButton--link AptoButton--primary activePage"
        data-page="10"
        type="button"
      >
        10
      </button>
    </li>
  </ul>
</div>
`;

exports[`AptoPager Renders no prev 1`] = `
<div
  class="AptoPager"
>
  <ul>
    <li
      class="AptoPager--item"
    >
      <button
        class="AptoButton AptoButton--link AptoButton--primary activePage"
        data-page="1"
        type="button"
      >
        1
      </button>
    </li>
    <li
      class="AptoPager--item"
    >
      <button
        class="AptoButton AptoButton--link AptoButton--primary"
        data-page="2"
        type="button"
      >
        2
      </button>
    </li>
    <li
      class="AptoPager--item"
    >
      <button
        class="AptoButton AptoButton--link AptoButton--primary"
        data-page="3"
        type="button"
      >
        3
      </button>
    </li>
    <li
      class="AptoPager--item"
    >
      <button
        class="AptoButton AptoButton--link AptoButton--primary"
        data-page="4"
        type="button"
      >
        4
      </button>
    </li>
    <li
      class="AptoPager--item"
    >
      <button
        class="AptoButton AptoButton--link AptoButton--primary"
        data-page="5"
        type="button"
      >
        5
      </button>
    </li>
    <li
      class="AptoPager--item AptoPager--next"
    >
      <button
        class="AptoButton AptoButton--link AptoButton--primary"
        data-page="2"
        type="button"
      >
        Next &gt;
      </button>
    </li>
  </ul>
</div>
`;

exports[`AptoPager Renders pages 1`] = `
<div
  class="AptoPager"
>
  <ul>
    <li
      class="AptoPager--item"
    >
      <button
        class="AptoButton AptoButton--link AptoButton--primary"
        data-page="1"
        type="button"
      >
        1
      </button>
    </li>
    <li
      class="AptoPager--item"
    >
      <button
        class="AptoButton AptoButton--link AptoButton--primary activePage"
        data-page="2"
        type="button"
      >
        2
      </button>
    </li>
    <li
      class="AptoPager--item"
    >
      <button
        class="AptoButton AptoButton--link AptoButton--primary"
        data-page="3"
        type="button"
      >
        3
      </button>
    </li>
    <li
      class="AptoPager--item"
    >
      <button
        class="AptoButton AptoButton--link AptoButton--primary"
        data-page="4"
        type="button"
      >
        4
      </button>
    </li>
    <li
      class="AptoPager--item"
    >
      <button
        class="AptoButton AptoButton--link AptoButton--primary"
        data-page="5"
        type="button"
      >
        5
      </button>
    </li>
  </ul>
</div>
`;
